### DISCUSSION TITLE
Suggestion for an addition to the Effect type

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#952] [Ideas]
Hey folks, I'd like to suggest that you add support for effect time-outs.

Say that I have an effect that makes an API call that could hang for a while. It would be nice if it could be automatically cancelled after some user-defined time has run without getting a response, possibly firing a user-defined action so that the reducer can respond to the original effect having timed out.

Many thanks!

**Comment by mbrandonw** [comment]
Hi @wtruppel, since `Effect` conforms to the `Publisher` protocol from Combine, you have access to all of the operators that Combine provides. In particular, there is a [`timeout`](https://developer.apple.com/documentation/combine/publisher/timeout(_:scheduler:options:customerror:)) operators that does just this.

You could use it like so:

```swift
case .buttonTapped:
  return environment.apiClient.fetch()
    .timeout(5, scheduler: environment.mainQueue)
    .catchToEffect()
```

You can also use the `customError` option of `timeout` to throw a specific error when the deadline is reached, and then use that in the reducer to execute logic for the timeout.

Also, if you use a [`TestScheduler`](https://github.com/pointfreeco/combine-schedulers#testscheduler) in your tests you will be able to write tests that prove your timeout logic works as you expect. 🙂

**Reply by ghost** [reply]
Ah, awesome! Many thanks for the reply. 👍
